File: python/citekit/mcp_server.py
"""CiteKit MCP Server — exposes getStructure and resolve tools.

Run with:
    python -m citekit.mcp_server

Or programmatically:
    from citekit.mcp_server import create_server
    server = create_server()
"""

from __future__ import annotations

import json
from pathlib import Path

from mcp.server import Server
from mcp.server.stdio import stdio_server
from mcp.types import TextContent, Tool

from citekit.client import CiteKitClient


def create_server(
    storage_dir: str = ".resource_maps",
    output_dir: str = ".citekit_output",
) -> Server:
    """Create an MCP server with CiteKit tools.

    Args:
        storage_dir: Directory where resource maps are stored.
        output_dir: Directory for resolved output files.
    """
    server = Server("citekit")
    client = CiteKitClient(storage_dir=storage_dir, output_dir=output_dir)

    @server.list_tools()
    async def list_tools() -> list[Tool]:
        return [
            Tool(
                name="listResources",
                description="List all available resource map IDs that have been ingested.",
                inputSchema={
                    "type": "object",
                    "properties": {},
                },
            ),
            Tool(
                name="getStructure",
                description=(
                    "Get the structured map of a resource. "
                    "Returns a JSON map with nodes, each pointing to a concept/section "
                    "and its physical location (pages, timestamps, regions). "
                    "Read the map to decide which node to resolve."
                ),
                inputSchema={
                    "type": "object",
                    "properties": {
                        "resource_id": {
                            "type": "string",
                            "description": "The resource ID to get the structure for.",
                        },
                    },
                    "required": ["resource_id"],
                },
            ),
            Tool(
                name="resolve",
                description=(
                    "Resolve a specific node from a resource map into extracted evidence. "
                    "For documents, this produces a mini-PDF with only the relevant pages. "
                    "For video/audio, this produces a clip of the relevant time segment. "
                    "For images, this produces a cropped region."
                ),
                inputSchema={
                    "type": "object",
                    "properties": {
                        "resource_id": {
                            "type": "string",
                            "description": "The resource ID.",
                        },
                        "node_id": {
                            "type": "string",
                            "description": "The node ID to resolve (e.g. 'derivatives.definition').",
                        },
                    },
                    "required": ["resource_id", "node_id"],
                },
            ),
        ]

    @server.call_tool()
    async def call_tool(name: str, arguments: dict) -> list[TextContent]:
        if name == "listResources":
            resource_ids = client.list_maps()
            return [TextContent(
                type="text",
                text=json.dumps({"resources": resource_ids}, indent=2),
            )]

        elif name == "getStructure":
            resource_id = arguments["resource_id"]
            try:
                structure = client.get_structure(resource_id)
                return [TextContent(
                    type="text",
                    text=json.dumps(structure, indent=2),
                )]
            except FileNotFoundError as e:
                return [TextContent(type="text", text=f"Error: {e}")]

        elif name == "resolve":
            resource_id = arguments["resource_id"]
            node_id = arguments["node_id"]
            try:
                evidence = client.resolve(resource_id, node_id)
                return [TextContent(
                    type="text",
                    text=json.dumps(evidence.model_dump(mode="json"), indent=2),
                )]
            except (FileNotFoundError, ValueError) as e:
                return [TextContent(type="text", text=f"Error: {e}")]

        else:
            return [TextContent(type="text", text=f"Unknown tool: {name}")]

    return server


async def main():
    """Run the CiteKit MCP server over stdio."""
    server = create_server()
    async with stdio_server() as (read_stream, write_stream):
        await server.run(read_stream, write_stream, server.create_initialization_options())


if __name__ == "__main__":
    import asyncio
    asyncio.run(main())

File: javascript/src/mcp-server.ts
/**
 * CiteKit MCP Server (JavaScript/TypeScript).
 *
 * Exposes listResources, getStructure, and resolve tools via MCP.
 * Reads maps from local JSON and delegates resolution to the Python SDK.
 *
 * Run with:
 *   npx tsx src/mcp-server.ts
 */

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import {
    ListToolsRequestSchema,
    CallToolRequestSchema,
} from "@modelcontextprotocol/sdk/types.js";
import { CiteKitClient } from "./client.js";
import { fileURLToPath } from "node:url";

function createServer(storageDir?: string, outputDir?: string): Server {
    const server = new Server(
        { name: "citekit", version: "0.1.7" },
        { capabilities: { tools: {} } }
    );

    const client = new CiteKitClient({ storageDir, outputDir });

    // ── List tools ──────────────────────────────────────────────────────────

    server.setRequestHandler(ListToolsRequestSchema, async () => ({
        tools: [
            {
                name: "listResources",
                description:
                    "List all available resource map IDs that have been ingested.",
                inputSchema: { type: "object" as const, properties: {} },
            },
            {
                name: "getStructure",
                description:
                    "Get the structured map of a resource. Returns nodes with concept/section info and physical locations.",
                inputSchema: {
                    type: "object" as const,
                    properties: {
                        resource_id: {
                            type: "string",
                            description: "The resource ID to get the structure for.",
                        },
                    },
                    required: ["resource_id"],
                },
            },
            {
                name: "resolve",
                description:
                    "Resolve a node into extracted evidence (mini-PDF, clip, or crop).",
                inputSchema: {
                    type: "object" as const,
                    properties: {
                        resource_id: {
                            type: "string",
                            description: "The resource ID.",
                        },
                        node_id: {
                            type: "string",
                            description: "The node ID to resolve.",
                        },
                        virtual: {
                            type: "boolean",
                            description: "If true, returns only metadata (timestamps/pages) without physical extraction. Useful for serverless/AI usage.",
                        },
                    },
                    required: ["resource_id", "node_id"],
                },
            },
            {
                name: "getNode",
                description:
                    "Get detailed metadata for a specific node without resolving it.",
                inputSchema: {
                    type: "object" as const,
                    properties: {
                        resource_id: {
                            type: "string",
                            description: "The resource ID.",
                        },
                        node_id: {
                            type: "string",
                            description: "The node ID.",
                        },
                    },
                    required: ["resource_id", "node_id"],
                },
            },
        ],
    }));

    // ── Call tools ──────────────────────────────────────────────────────────

    server.setRequestHandler(CallToolRequestSchema, async (request) => {
        const { name, arguments: args } = request.params;

        try {
            if (name === "listResources") {
                const resources = client.listMaps();
                return {
                    content: [
                        {
                            type: "text" as const,
                            text: JSON.stringify({ resources }, null, 2),
                        },
                    ],
                };
            }

            if (name === "getStructure") {
                const resourceId = (args as Record<string, string>).resource_id;
                const structure = client.getStructure(resourceId);
                return {
                    content: [
                        {
                            type: "text" as const,
                            text: JSON.stringify(structure, null, 2),
                        },
                    ],
                };
            }

            if (name === "getNode") {
                const { resource_id, node_id } = args as Record<string, string>;
                const structure = client.getStructure(resource_id);

                function findNode(nodes: any[], targetId: string): any {
                    for (const node of nodes) {
                        if (node.id === targetId) return node;
                        if (node.children) {
                            const found = findNode(node.children, targetId);
                            if (found) return found;
                        }
                    }
                    return null;
                }

                const node = findNode(structure.nodes, node_id);
                if (!node) {
                    throw new Error(`Node '${node_id}' not found in resource '${resource_id}'`);
                }

                return {
                    content: [
                        {
                            type: "text" as const,
                            text: JSON.stringify(node, null, 2),
                        },
                    ],
                };
            }

            if (name === "resolve") {
                const { resource_id, node_id, virtual } = args as any;
                const evidence = await client.resolve(resource_id, node_id, { virtual: !!virtual });
                return {
                    content: [
                        {
                            type: "text" as const,
                            text: JSON.stringify(evidence, null, 2),
                        },
                    ],
                };
            }

            return {
                content: [{ type: "text" as const, text: `Unknown tool: ${name}` }],
                isError: true,
            };
        } catch (error) {
            return {
                content: [
                    {
                        type: "text" as const,
                        text: `Error: ${error instanceof Error ? error.message : String(error)}`,
                    },
                ],
                isError: true,
            };
        }
    });

    return server;
}

// ── Main ────────────────────────────────────────────────────────────────────

export async function runMcpServer() {
    const server = createServer();
    const transport = new StdioServerTransport();
    await server.connect(transport);
}

// Only run if called directly
if (fileURLToPath(import.meta.url) === process.argv[1]) {
    runMcpServer().catch(console.error);
}
